LLVM: support -fuse-ld=lld linker on qemu_x86.

**Describe the bug**
when tried to **use LLVM built-in lld linker to build zephyr case**, it met a **link error "section 'boot.text' will not fit in region 'RAM': overflowed by 18446744072631615488 bytes"**


**To Reproduce**
Steps to reproduce the behavior:
1. set LLVM toolchain:
     export ZEPHYR_TOOLCHAIN_VARIANT=llvm
     export LLVM_TOOLCHAIN_PATH=~/llvm/llvm-project/build_2
     (notice: **(the LLVM version must be after this commit https://github.com/llvm/llvm-project/commit/2bf06d9345caeb26520be8e830c092683bbdf0f7, otherwise, it couldn't support parsing quoted symbol like "__start" which is used for zephyr entry point)**)
2. apply this patch, it's used for forcing llvm to use lld and solving option compatibility between ld and lld.
```
diff --git a/Kconfig.zephyr b/Kconfig.zephyr
index d2d3e0a8db..349a98da7b 100644
--- a/Kconfig.zephyr
+++ b/Kconfig.zephyr
@@ -162,6 +162,12 @@ config CMAKE_LINKER_GENERATOR

 endchoice

+config LLVM_USE_BUILTIN_LLD
+       bool "LLVM uses built-in lld linker"
+       default n
+       help
+         Force LLVM to use built-in lld linker.
+
 config HAVE_CUSTOM_LINKER_SCRIPT
        bool "Custom linker script provided"
        help
diff --git a/cmake/linker/ld/target_base.cmake b/cmake/linker/ld/target_base.cmake
index 1fec5ca96f..69a008b17a 100644
--- a/cmake/linker/ld/target_base.cmake
+++ b/cmake/linker/ld/target_base.cmake
@@ -24,8 +24,15 @@ macro(toolchain_ld_base)
   # in descending order to minimize padding between these symbols.
   zephyr_ld_option_ifdef(
     CONFIG_LINKER_SORT_BY_ALIGNMENT
-    ${LINKERFLAGPREFIX},--sort-common=descending
     ${LINKERFLAGPREFIX},--sort-section=alignment
   )
+  # --sort-common option of lld is not compatible with ld, so remove
+  # it when using lld.
+  if(NOT CONFIG_LLVM_USE_BUILTIN_LLD)
+    zephyr_ld_option_ifdef(
+      CONFIG_LINKER_SORT_BY_ALIGNMENT
+      ${LINKERFLAGPREFIX},--sort-common=descending
+    )
+  endif()

 endmacro()
diff --git a/cmake/linker/lld/target.cmake b/cmake/linker/lld/target.cmake
index e9ac00ff96..376a26b680 100644
--- a/cmake/linker/lld/target.cmake
+++ b/cmake/linker/lld/target.cmake
@@ -11,6 +11,9 @@ macro(configure_linker_script linker_script_gen linker_pass_define)
   set(extra_dependencies ${ARGN})
   set(template_script_defines ${linker_pass_define})
   list(TRANSFORM template_script_defines PREPEND "-D")
+  if(CONFIG_LLVM_USE_BUILTIN_LLD)
+    set_property(GLOBAL APPEND PROPERTY PROPERTY_LINKER_SCRIPT_DEFINES -D__LLVM_LINKER_CMD__)
+  endif()

   # Different generators deal with depfiles differently.
   if(CMAKE_GENERATOR STREQUAL "Unix Makefiles")
diff --git a/cmake/linker/lld/target_baremetal.cmake b/cmake/linker/lld/target_baremetal.cmake
index 76965ed81b..0b315828cb 100644
--- a/cmake/linker/lld/target_baremetal.cmake
+++ b/cmake/linker/lld/target_baremetal.cmake
@@ -11,6 +11,12 @@ macro(toolchain_ld_baremetal)
     ${LINKERFLAGPREFIX},-X
     ${LINKERFLAGPREFIX},-N
   )
+  # Force LLVM to use built-in linker lld
+  if(CONFIG_LLVM_USE_BUILTIN_LLD)
+    zephyr_ld_options(
+      -fuse-ld=lld
+    )
+  endif()

   # Funny thing is if this is set to =error, some architectures will
   # skip this flag even though the compiler flag check passes
diff --git a/include/linker/linker-tool.h b/include/linker/linker-tool.h
index 352cc5f60e..2c1ea7508c 100644
--- a/include/linker/linker-tool.h
+++ b/include/linker/linker-tool.h
@@ -17,6 +17,13 @@

 #if defined(_LINKER)
 #if defined(__GCC_LINKER_CMD__)
+/* ALIGN_WITH_INPUT is not supported by LLVM lld,
+ * it needs to be replaced by ALIGN(8).
+ */
+#if defined(__LLVM_LINKER_CMD__)
+#define ALIGN_WITH_INPUT ALIGN(8)
+#endif
+
 #include <linker/linker-tool-gcc.h>
 #elif defined(__MWDT_LINKER_CMD__)
 #include <linker/linker-tool-mwdt.h>
```

    
3.  west build -b qemu_x86 -p auto samples/hello_world/ -DCONFIG_LLVM_USE_BUILTIN_LLD=y
4. See error

**Logs and console output**
[83/97] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /home/chenpeng/zephyrproject/zephyr/build/zephyr/zephyr_pre0.map
: && ccache **/home/chenpeng/llvm/llvm-project/build_2/bin/clang --target=i686-pc-none-elf**   zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/x86/core/offsets/offsets.c.obj  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/home/chenpeng/zephyrproject/zephyr/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/arch/x86/core/libarch__x86__core.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/subsys/random/libsubsys__random.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/usr/lib64/gcc/x86_64-redhat-linux/9/32"  -L/home/chenpeng/zephyrproject/zephyr/build/zephyr  -lgcc  -no-pie  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  **-fuse-ld=lld**  -Wl,--orphan-handling=warn && cd /home/chenpeng/zephyrproject/zephyr/build/zephyr && /home/chenpeng/.local/lib/python3.7/site-packages/cmake/data/bin/cmake -E echo
clang-13: warning: argument unused during compilation: '-no-pie' [-Wunused-command-line-argument]
ld.lld: warning: <internal>:(.eh_frame) is being placed in '.eh_frame'
ld.lld: warning: <internal>:(.symtab) is being placed in '.symtab'
ld.lld: warning: <internal>:(.shstrtab) is being placed in '.shstrtab'
ld.lld: warning: <internal>:(.strtab) is being placed in '.strtab'
**ld.lld: error: no memory region specified for section '.eh_frame'
ld.lld: error: zephyr/linker_zephyr_pre0.cmd:101: unable to move location counter backward for: text
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705357312 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705357312 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705357473 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705357754 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705358093 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705358098 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705358121 bytes**
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705358134 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705358149 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705358191 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705358239 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705358523 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705358605 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705359498 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705359722 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705359870 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705359900 bytes
ld.lld: error: section 'text' will not fit in region 'RAM': overflowed by 18446744073705359914 bytes
ld.lld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
collect2: error: ld returned 1 exit status
clang-13: error: linker (via gcc) command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /home/chenpeng/.local/bin/cmake --build /home/chenpeng/zephyrproject/zephyr/build


Part of code of liner script linker_zephyr_pre0.cmd:
```
MEMORY
    {
    RAM (wx) : ORIGIN = (0 + 0x100000), LENGTH = (4194304 - 0x100000)
    IDT_LIST : ORIGIN = 0xFFFF1000, LENGTH = 2K

    }
  OUTPUT_FORMAT("elf32-i386", "elf32-i386", "elf32-i386")
  OUTPUT_ARCH("i386")
epoint = ((__start) - ((0x0 + 0x100000) - (0x0 + 0x100000)));
ENTRY(epoint)
 text :
 {
 __text_region_start = .;
 z_mapped_start = .;
. = 0x0;
. = ALIGN(4);
 KEEP(*(.text.__start))
 *(.text)
 *(".text.*")
 *(.gnu.linkonce.t.*)
 *(.init)
 *(.fini)
 *(.eini)
 . = ALIGN(0x1000);
 } > RAM
```
**Environment (please complete the following information):**
 - OS: Fedora
 - Toolchain: LLVM 
 - Zephyr Commit: 0e96f59e8365411dc9493

--- a/Kconfig.zephyr
+++ b/Kconfig.zephyr
@@ -162,6 +162,13 @@ config CMAKE_LINKER_GENERATOR
 
 endchoice
 
+config LLVM_USE_LD
+	bool "LLVM use ld linker"
+	depends on "${ZEPHYR_TOOLCHAIN_VARIANT}" = "llvm"
+	default y
+	help
+	  Use binutils ld linker instead of LLVM built-in lld linker.
+
 config HAVE_CUSTOM_LINKER_SCRIPT
 	bool "Custom linker script provided"
 	help

--- a/cmake/linker/lld/target.cmake
+++ b/cmake/linker/lld/target.cmake
@@ -107,7 +107,7 @@ endfunction(toolchain_ld_link_elf)
 
 
 # Load toolchain_ld-family macros
-include(${ZEPHYR_BASE}/cmake/linker/ld/target_base.cmake)
+include(${ZEPHYR_BASE}/cmake/linker/${LINKER}/target_base.cmake)
 include(${ZEPHYR_BASE}/cmake/linker/${LINKER}/target_baremetal.cmake)
 include(${ZEPHYR_BASE}/cmake/linker/ld/target_cpp.cmake)
 include(${ZEPHYR_BASE}/cmake/linker/ld/target_relocation.cmake)

--- a/cmake/linker/lld/target_baremetal.cmake
+++ b/cmake/linker/lld/target_baremetal.cmake
@@ -4,13 +4,20 @@
 
 macro(toolchain_ld_baremetal)
 
-  # LINKERFLAGPREFIX comes from linker/ld/target.cmake
+  # LINKERFLAGPREFIX comes from linker/lld/target.cmake
   zephyr_ld_options(
     -nostdlib
     -static
     ${LINKERFLAGPREFIX},-X
     ${LINKERFLAGPREFIX},-N
   )
+
+  # Force LLVM to use built-in lld linker
+  if(NOT CONFIG_LLVM_USE_LD)
+    zephyr_ld_options(
+      -fuse-ld=lld
+  )
+  endif()
 
   # Funny thing is if this is set to =error, some architectures will
   # skip this flag even though the compiler flag check passes

--- a/cmake/linker/lld/target_base.cmake
+++ b/cmake/linker/lld/target_base.cmake
@@ -1,0 +1,28 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# See root CMakeLists.txt for description and expectations of these macros
+
+macro(toolchain_ld_base)
+
+  if(NOT PROPERTY_LINKER_SCRIPT_DEFINES)
+    set_property(GLOBAL PROPERTY PROPERTY_LINKER_SCRIPT_DEFINES -D__GCC_LINKER_CMD__)
+  endif()
+
+  # TOOLCHAIN_LD_FLAGS comes from compiler/clang/target.cmake
+  # LINKERFLAGPREFIX comes from linker/lld/target.cmake
+  zephyr_ld_options(
+    ${TOOLCHAIN_LD_FLAGS}
+  )
+
+  zephyr_ld_options(
+    ${LINKERFLAGPREFIX},--gc-sections
+    ${LINKERFLAGPREFIX},--build-id=none
+  )
+
+  # Sort each input section by alignment.
+  zephyr_ld_option_ifdef(
+    CONFIG_LINKER_SORT_BY_ALIGNMENT
+    ${LINKERFLAGPREFIX},--sort-section=alignment
+  )
+
+endmacro()

--- a/cmake/toolchain/llvm/generic.cmake
+++ b/cmake/toolchain/llvm/generic.cmake
@@ -9,7 +9,11 @@ endif()
 set(LLVM_TOOLCHAIN_PATH ${CLANG_ROOT_DIR} CACHE PATH "clang install directory")
 
 set(COMPILER clang)
-set(LINKER lld)
+if(CONFIG_LLVM_USE_LD)
+  set(LINKER ld)
+else()
+  set(LINKER lld)
+endif()
 set(BINTOOLS llvm)
 
 if("${ARCH}" STREQUAL "arm")

--- a/include/zephyr/arch/x86/ia32/linker.ld
+++ b/include/zephyr/arch/x86/ia32/linker.ld
@@ -313,6 +313,15 @@ SECTIONS
 
 	/* Needs KEEP() as ENTRY() is given a physical address */
 	KEEP(*(.text.__start))
+/*
+ * We need these sections to extract interrupt information, but they
+ * will be removed with "--gc-sections" by LLVM lld, so add keep
+ * command to save them.
+ */
+#ifndef CONFIG_LLVM_USE_LD
+	KEEP(*(.text.irqstubs))
+	KEEP(*(".gnu.linkonce.t.exc_*_stub"))
+#endif
 	*(.text)
 	*(".text.*")
 	*(.gnu.linkonce.t.*)
@@ -532,7 +541,22 @@ SECTIONS
 #include <zephyr/linker/debug-sections.ld>
 
 	/DISCARD/ : { *(.note.GNU-stack) }
+/*
+ * eh_frame section won't be removed even with "--gc-sections" by LLVM lld.
+ */
+#if !defined(CONFIG_EXCEPTIONS)
+	/DISCARD/ : { *(.eh_frame) }
+#endif
 
+/*
+ * The sections below are still treated as warnings
+ * with "--orphan-handling=warn" by LLVM lld.
+ */
+#if !defined(CONFIG_LLVM_USE_LD)
+	.symtab 0 : { *(.symtab) }
+	.strtab 0 : { *(.strtab) }
+	.shstrtab 0 : { *(.shstrtab) }
+#endif
 	}
 
 #ifdef CONFIG_XIP

--- a/include/zephyr/arch/x86/intel64/linker.ld
+++ b/include/zephyr/arch/x86/intel64/linker.ld
@@ -217,4 +217,21 @@ SECTIONS
 	*(.rel.*)
 	*(.rela.*)
 	}
+/*
+ * eh_frame section won't be removed even with "--gc-sections" by LLVM lld.
+ */
+#if !defined(CONFIG_EXCEPTIONS)
+	/DISCARD/ : { *(.eh_frame) }
+#endif
+
+/*
+ * The sections below are still treated as warnings
+ * with "--orphan-handling=warn" by LLVM lld.
+ */
+#if !defined(CONFIG_LLVM_USE_LD)
+	.symtab 0 : { *(.symtab) }
+	.strtab 0 : { *(.strtab) }
+	.shstrtab 0 : { *(.shstrtab) }
+#endif
+
 }

